### PR TITLE
Validate IPv6 SRH hop addresses

### PR DESCRIPTION
**Pull request by nvt**
Prevent IPv6 SRH processing when the next hop is our own address, or an invalid address for forwarding. This prevents a bug that causes deep recursive calls that occur because of the way that uIP processes packets.

### CODE REVIEW

**Review by arurke — APPROVED** [review]

**Comment by simonduq** [comment]
Hi!

This was my mistake, very sorry about it, re-opening this PR now.

What happened is the following: I was switching the base branch for this repository https://github.com/wittra/contiki-ng from develop to wittra. But accidentally, I made the change on the wrong repo (this repo) and instead of switching the base branch I renamed it. And somehow github deleted develop and closed all PRs...

Many apologies for this mishap 🙏; I haven't contributed in a while.. but now at least everybody got some notification from me :p

**Comment by joakimeriksson** [comment]
Maybe we should add a validation loop in the extension header processing code?
RFC 6554 says: "  To detect loops in the SRH, a router MUST determine if the SRH
   includes multiple addresses assigned to any interface on that router.
   If such addresses appear more than once and are separated by at least
   one address not assigned to that router, the router MUST drop the
   packet"

**Comment by Scepticz** [comment]
Hi there,

are there any updates on the fix? There is an advisory draft CVE-2023-29001. The preliminary title seems to be "Uncontrolled recursion due to insufficient validation of the IPv6 source routing header". It would be great to have that published in the process as well. :-)

Best
Tobi

**Comment by nvt** [comment]
Updated with an SRH hop loop check.

File: os/net/routing/rpl-classic/rpl-ext-header.c [hunk 1]
```diff
@@ -209,6 +209,72 @@ rpl_ext_header_srh_get_next_hop(uip_ipaddr_t *ipaddr)
...
 /*---------------------------------------------------------------------------*/
+#if RPL_WITH_NON_STORING
+static bool
+srh_is_valid(struct uip_routing_hdr *rh_header,
+             struct uip_rpl_srh_hdr *srh_header)
+{
+  uip_ipaddr_t hop_addr;
+  uip_ipaddr_copy(&hop_addr, &UIP_IP_BUF->destipaddr);
+
+  uint8_t  segments_left = rh_header->seg_left;
```
Comment (pjonsson) [inline os/net/routing/rpl-classic/rpl-ext-header.c:220]: One space too much added after `uint8_t'.

File: os/net/routing/rpl-lite/rpl-ext-header.c [hunk 2]
```diff
@@ -87,6 +87,70 @@ rpl_ext_header_srh_get_next_hop(uip_ipaddr_t *ipaddr)
...
 }
 /*---------------------------------------------------------------------------*/
+static bool
+srh_is_valid(struct uip_routing_hdr *rh_header,
+             struct uip_rpl_srh_hdr *srh_header)
+{
+  uip_ipaddr_t hop_addr;
+  uip_ipaddr_copy(&hop_addr, &UIP_IP_BUF->destipaddr);
+
+  uint8_t  segments_left = rh_header->seg_left;
```
Comment (pjonsson) [inline os/net/routing/rpl-lite/rpl-ext-header.c:97]: One space too much added after `uint8_t`.

**Review by pjonsson — APPROVED** [review]

**Pull request merged**